include/grpc/support/alloc.h

```diff
@@ -52,15 +45,6 @@ GPRAPI void* gpr_malloc_aligned(size_t size, size_t alignment);
 /** free memory allocated by gpr_malloc_aligned */
 GPRAPI void gpr_free_aligned(void* ptr);
```

Comment:
If we're getting rid of the mechanism to override the behavior of these functions, do we still need them?  Could we instead just change our code to use `malloc()`, `free()`, et al directly?

Reply:
Since the core library allocates buffers and transfers ownership to the app, it needs to continue to provide deallocation functions, doesn't it? The app needs to be able to free the gRPC-allocated buffer, but the app's `free` and the core library's `free` may not be the same when the C runtime is statically linked (as is often the case when building gRPC for Windows). `gpr_free` means that the calling code can get back to the right heap.

Reply:
As chwarr said, it's still required to keep them in order to match `malloc` and `free` in case where the ownership of buffer allocated by application is passed to the gRPC library and vice versa. But, most of `grpc_malloc` and `gpr_free` can be replaced with the regular ones if its ownership doesn't transfer.

Reply:
Are there actually cases where C-core allocates memory and returns it to the application, which is expected to call `gpr_free()` on it?  For example, for message payloads, core returns a `grpc_byte_buffer` and expects the application to call `grpc_byte_buffer_destroy()` on it, so the application never needs to call `gpr_free()` directly.

I have not actually looked through the C-core API, so there may be cases where we are doing this.  But please check the API to make sure that there are actually such cases.  If there are not, then we don't actually need to expose these functions.

Reply:
I haven't looked through all cases involved and definitely it might be valuable to see that too. We might come up with more narrow API such as `grpc_alloc_metadata` and `gprc_free_metadata` if only metadata will go in and out across the API border instead of having general `gpr_malloc` and `gpr_free`.

How about merging this first and working on `malloc` and `free`later? It seems that there are a few of usages even in internal repository.

Reply:
The following non-exhaustive list of APIs in the "basic" core API allocate and require the caller/app the free via `gpr_free`:

* [`grpc_call_get_peer`][grpc_call_get_peer]
* [`grpc_ssl_roots_override_result`][grpc_ssl_roots_override_result]
* [`grpc_slice_to_c_string`][grpc_slice_to_c_string]

There are a few more in more obscure parts of the library, like `gpr_strdup` and `gpr_asprintf`.

[grpc_call_get_peer]: https://github.com/grpc/grpc/blob/834f0c780900d3ce6e917c23f6dacb4f6d8e4b56/include/grpc/grpc.h#L269-L277
[grpc_ssl_roots_override_result]: https://github.com/grpc/grpc/blob/0362df725f1ed9db3cb3e50456e2ad59cb4a1c1b/include/grpc/grpc_security.h#L138-L145
[grpc_slice_to_c_string]: https://github.com/grpc/grpc/blob/f8b1e6cb4401c062eba1f6edbffcbf438260fadd/include/grpc/slice.h#L164-L166

Reply:
Okay, I'm fine with doing this in a separate change.  Let's just leave a TODO about it for now.

Reply:
Thank you chwarr@. From your list, the idea to have separate free methods for each caller doesn't seem valid since there are a couple of functions. markrdroth@ I don't think that we can change this in this way.

Reply:
I don't see any reason that this is impossible.  For example, for `grpc_call_get_peer()`, we could change its return type to be a custom peer object with its own dtor.  For `grpc_slice_to_c_string()`, we could investigate changing it to not duplicate the string but instead return a pointer to the existing string and make it the caller's responsibility to duplicate it.  We just have to look at each of these cases on a case-by-case basis.

In any case, I agree that this is future work; it's not something we need to address in this PR.

Reply:
Oh I see. You're thinking of using C++ for Core interface to address this. And then it would be feasible. But we need to discuss which way is better; i) keeping gpr_malloc and gpr_free to address memory ownership transfer, ii) adding dedicate alloc/dealloc functions per usage or iii) changing it to C++ API with a destructor. Let's do this later.

Reply:
The change I am proposing here does not have to be in C++.  When I said "with its own dtor", I didn't necessarily mean that it's a C++ class with a dtor; it could equally be a C type with a destruction method.  For example, we could do something like this:

```
typedef struct grpc_peer grpc_peer;
// Gets the name of the peer from the peer object.
// Caller does not take ownership of the returned string.
const char* grpc_peer_get_name(grpc_peer* peer);
// Destroys the peer object.
void grpc_peer_destroy(grpc_peer* peer);

// Returns the peer of the call.
grpc_peer* grpc_call_get_peer(grpc_call* call);
```

In any case, I agree that we can discuss this later.